platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java

```diff
@@ -0,0 +1,178 @@
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.configurations;
+
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.MultimapBuilder;
+import org.gradle.api.GradleException;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.capabilities.Capability;
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.internal.component.external.model.ImmutableCapabilities;
+import org.gradle.internal.component.external.model.ProjectDerivedCapability;
+import org.gradle.internal.deprecation.DocumentedFailure;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Static utility to verify a set of variants each have a unique identity in terms of attributes and capabilities.
+ */
+public class VariantIdentityUniquenessVerifier {
+
+    /**
+     * Build a report of all possible variant uniqueness failures for the given configurations.
+     */
+    public static VerificationReport buildReport(ConfigurationsProvider configurations) {
+        ListMultimap<VariantIdentity, ConfigurationInternal> byIdentity =
+            MultimapBuilder.linkedHashKeys().arrayListValues().build();
+
+        configurations.visitAll(configuration -> {
+            if (!mustHaveUniqueAttributes(configuration)) {
+                return;
+            }
+
+            byIdentity.put(VariantIdentity.from(configuration), configuration);
+        });
+
+        return new VerificationReport(byIdentity);
+    }
+
+    /**
+     * Consumable, non-resolvable, non-default configurations with attributes must have unique attributes.
+     */
+    private static boolean mustHaveUniqueAttributes(Configuration configuration) {
+        return configuration.isCanBeConsumed() &&
+            !configuration.isCanBeResolved() &&
+            !Dependency.DEFAULT_CONFIGURATION.equals(configuration.getName()) &&
+            !configuration.getAttributes().isEmpty();
+    }
+
+    /**
+     * A report tracking all possible variant uniqueness failures for a component.
+     */
+    public static class VerificationReport {
+
+        private final ListMultimap<VariantIdentity, ConfigurationInternal> byIdentity;
+
+        private VerificationReport(ListMultimap<VariantIdentity, ConfigurationInternal> byIdentity) {
+            this.byIdentity = byIdentity;
+        }
+
+        /**
+         * Get a failure that only checks variant uniqueness for the given configuration.
+         */
+        @Nullable
+        public GradleException failureFor(ConfigurationInternal configuration, boolean withTaskAdvice) {
+            List<ConfigurationInternal> collisions =
+                byIdentity.get(VariantIdentity.from(configuration)).stream()
+                    .filter(it -> it != configuration)
```

Comment:
❗ I think this find a matching configuration by name instead of instance.  Actually maybe the parameter should just be the name, too?  

I don't see why we should need to keep around the entire configuration to query the failures here, or why using a copy/clone with the same name shouldn't report the same failures.  At least while configuration live in a global container with globally unique names.

Reply:
I've switched to a name comparison. However I've left the configuration as a parameter since we need to call `VariantIdentity.from(configuration)` and I'd rather not expose `VariantIdentity` right now